orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/util/PackageInfo.groovy

```diff
@@ -90,62 +90,72 @@ class PackageInfo {
       }
-    }
 
-    if (buildArtifact) {
-      packageName = extractPackageName(buildArtifact, fileExtension)
-      if (extractVersion) {
-        packageVersion = extractPackageVersion(buildArtifact, prefix, fileExtension)
+      String packageName
+      String packageVersion
+
+      if (triggerArtifact) {
+        packageName = extractPackageName(triggerArtifact, fileExtension)
+        if (extractVersion) {
+          packageVersion = extractPackageVersion(triggerArtifact, prefix, fileExtension)
+        }
       }
-    }
 
-    if (packageVersion) {
-      request.put('packageVersion', packageVersion)
-    }
+      if (buildArtifact) {
+        packageName = extractPackageName(buildArtifact, fileExtension)
+        if (extractVersion) {
+          packageVersion = extractPackageVersion(buildArtifact, prefix, fileExtension)
+        }
+      }
 
-    if (packageName) {
-      request.put('package', packageName)
+      if (packageVersion) {
+        request.put('packageVersion', packageVersion)
+      }
+
+      // If the package cannot be found it's passed down as is
+      if (!triggerArtifact && !buildArtifact) {
+        packageName = requestPackage
+      }
 
-      if (extractBuildDetails) {
-        def buildInfoUrl = buildArtifact ? buildInfo?.url : trigger?.buildInfo?.url
-        def buildInfoUrlParts = parseBuildInfoUrl(buildInfoUrl)
+      if (packageName) {
 
-        if (buildInfoUrlParts?.size == 3) {
-          request.put('buildHost', buildInfoUrlParts[0].toString())
-          request.put('job', buildInfoUrlParts[1].toString())
-          request.put('buildNumber', buildInfoUrlParts[2].toString())
-        }
+        requestPackages[index] = packageName
 
-        def commitHash = buildArtifact ? extractCommitHash(buildInfo) : extractCommitHash(trigger?.buildInfo)
+        if (extractBuildDetails) {
+          def buildInfoUrl = buildArtifact ? buildInfo?.url : trigger?.buildInfo?.url
+          def buildInfoUrlParts = parseBuildInfoUrl(buildInfoUrl)
 
-        if (commitHash) {
-          request.put('commitHash', commitHash)
+          if (buildInfoUrlParts?.size == 3) {
+            request.put('buildHost', buildInfoUrlParts[0].toString())
+            request.put('job', buildInfoUrlParts[1].toString())
+            request.put('buildNumber', buildInfoUrlParts[2].toString())
+          }
+
+          def commitHash = buildArtifact ? extractCommitHash(buildInfo) : extractCommitHash(trigger?.buildInfo)
+
+          if (commitHash) {
+            request.put('commitHash', commitHash)
+          }
         }
       }
-
-      return request
     }
 
-    throw new IllegalStateException("Unable to find deployable artifact starting with ${prefix} and ending with ${fileExtension} in ${buildArtifacts} and ${triggerArtifacts}. Make sure your deb package file name complies with the naming convention: name_version-release_arch.")
+    request.replace('package', requestPackages.join(" "))
```

Comment:
Seems like there are 2 changes in this PR:
1) Allow the package that will be matched to a trigger/build artifact to occupy slots other than just sub 0.
2) Remove the requirement that a matching artifact in fact exists in the context.

Change (1) LGTM. It reads to me like we can make change (1) without change (2). Can you just track whether or not a match was made (I think this is what `if (packageName)` already does) and then throw that exception if no match was made over all of the specified packages? And if a match was made, then the augmented request is returned.